gossipsub.go

```diff
@@ -1617,6 +1693,13 @@ func (gs *GossipSubRouter) clearIHaveCounters() {
 	}
 }
 
+func (gs *GossipSubRouter) clearIDontWantCounters() {
```

Comment:
Thats too aggressive and can clear the counters at unfortunate times.
Lets keep them for heartbeat or two, maybe 3 to match the cache sliding window.

Reply:
I don't understand this. When is the unfortunate time?

This only clear the counters not the cache that keeps the message ids from IDONTWANTs. So technically, the slower you clear the cache, the more you prevent the DoS attack. The faster you clear the cache, the more efficiency you have (because you process more IDONTWANTs).

Reply:
Lets say you just got an IDW, heartbrrat clear, and then the message arrives... and you send it even though it is not wanted.

Reply:
> and you send it even though it is not wanted.

This is not true.

The heartbeat clear only clears the counter, not the cache. The IDONTWANT is still in the cache so the message is not forwarded. In other words, the heartbeat clear clears `peerdontwant`, while only `unwanted` is used to decide whether to forward it or not.

Reply:
when is the cache cleared?

Reply:
> when is the cache cleared?

Currently, it's not cleared at all. It's currently not specified in the spec (even if it probably should be).

As I read from the spec, `max_idontwant_messages` is only used for limiting the number of IDONTWANTs to be processed per heartbeat per peer.

If the intention is really to limit the number of IDONTWANTs the node should remember in the cache to save the memory, I think the current spec doesn't make any sense because you should do an LRU with a max size instead of doing the heartbeat thing.

Reply:
we need to clear the cache, otherwise it is compleltely broken.

Reply:
keep a counter of how many heartbeats each IDW is kept for, and clear i after 3.

Reply:
> keep a counter of how many heartbeats each IDW is kept for, and clear it after 3.

That makes sense. I will do it, but I think it's considered go-libp2p specific and out of the scope of the spec, right?

Reply:
implementation detail perhaps.

Reply:
Wait. If the purpose is to just limit the memory used, why not just define a parameter for the maximum number of IDONTWANTs to keep in the cache per peer?

If the number in the cache grows beyond the limit, just remove the least recently seen IDONTWANT from the cache before adding a new one.

I think this is simpler than dealing with the heartbeats.

Reply:
you misunderatand.

We need to keep the IDW for some heartbeats and then clear them.

It is both for controlling memory use and for correctness.

Reply:
reagardless, an LRA (least recentlybadded) cache for IDW could work and keep memory bounded and it is possibly simpler.

However we need to consider cache sizes carefully.

Reply:
and the problem here is a sybil attack, and we get to keep all the garbage indefinitely.

So no, this is not the correct solution.

We need to explicitly clean old IDWs.

Reply:
In fact we should go one step further and use a global limit for IDWs across all peers in the cache.

Reply:
> It is both for controlling memory use and for correctness.

Alright, got it. Thanks.